plutus-ledger-api/src/PlutusLedgerApi/Common/SerialisedScript.hs

```diff
@@ -155,6 +159,17 @@ serialiseUPLC =
   -- need to be careful about introducing a working version
   toShort . BSL.toStrict . serialise . SerialiseViaFlat . UPLC.UnrestrictedProgram
 
+{- | Hash a 'SerialisedScript' with the given version prefix. Each prefix corresponds
+to a specific version of the Plutus language.
+
+As of PlutusV3, the Plutus language versions and corresponding prefixes are as follows:
+  PlutusV1 -> 0x1
+  PlutusV2 -> 0x2
+  PlutusV3 -> 0x3
+-}
+hashScriptWithPrefix :: Word8 -> SerialisedScript -> BS.ByteString
```

Comment:
How about:
```haskell
-- | Hash a 'SerialisedScript' with the Plutus language version 1 prefix.
hashScriptV1 :: SerialisedScript -> BS.ByteString
hashScriptV1 = hashScriptWithPrefix 0x1 

-- | Hash a 'SerialisedScript' with the Plutus language version 2 prefix.
hashScriptV2 :: SerialisedScript -> BS.ByteString
hashScriptV2 = hashScriptWithPrefix 0x2

-- | Hash a 'SerialisedScript' with the Plutus language version 3 prefix.
hashScriptV3 :: SerialisedScript -> BS.ByteString
hashScriptV3 = hashScriptWithPrefix 0x3
```

Reply:
Also replace [this](https://github.com/IntersectMBO/plutus/blob/ed5cd39a57584f3d05258199a3903b6e5caf2d74/plutus-tx/src/PlutusTx/Blueprint/Validator.hs#L58-L59) with the new functions.

Reply:
> This approach also makes invalid prefixes un-representable. I wouldn't stop here though:
`PlutusLedgerApi.Common.hashScriptV1`  move to `PlutusLedgerApi.V1.hashScript`
`PlutusLedgerApi.Common.hashScriptV2`  move to `PlutusLedgerApi.V2.hashScript`
`PlutusLedgerApi.Common.hashScriptV3`  move to `PlutusLedgerApi.V3.hashScript`
wdyt?

Yes and we probably don't need `hashScriptWithPrefix` at all, you can just copy-paste the body of that definition within each `V*` file. I'm just nitpicking though, feel free to ignore.

Reply:
> Also replace [this](https://github.com/IntersectMBO/plutus/blob/ed5cd39a57584f3d05258199a3903b6e5caf2d74/plutus-tx/src/PlutusTx/Blueprint/Validator.hs#L58-L59) with the new functions.

This introduces a cyclic dependency between `plutus-tx` and `plutus-ledger-api`.

Reply:
Personally I'd do what @effectfully  suggests: inline the `hasScriptWithPrefix` into V1/V2/V3 `hashScript`s.

Reply:
The reason I want to keep it is because `hashScriptWithPrefix` as an export is useful to me in creating and testing new language versions that I use in Midgard (optimistic rollup building on Cardano), and it will also be useful for the Hydra team for the same reason.

Reply:
> This introduces a cyclic dependency between plutus-tx and plutus-ledger-api.

Right, so this should really exist in a `plutus-script-utils` package, which users should depend on, rather than `plutus-ledger-api` which is intended to be stuff the ledger needs. This is the reason why a number of functions similar to this one was removed from `plutus-ledger-api` in the first place. This nonetheless can be future work.